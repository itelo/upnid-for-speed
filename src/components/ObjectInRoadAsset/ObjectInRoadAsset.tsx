import React from "react";
import { motion } from "framer-motion";
import rockImg from "../../assets/rock.png";
import lightningImg from "../../assets/lightning-bolt.png";
import styled from "styled-components";

type ObjectData = {
  key: string;
  type: "rock" | "lightning";
  y: number;
};

type ObjectInRoadAssetProps = {
  size: number;
  matrix: { x: number; y: number; scale: number }[][];
  onAnimationComplete: (obj: ObjectData, index: number) => () => void;
  index: number;
  obj: ObjectData;
} & React.ImgHTMLAttributes<HTMLImageElement>;

const ObjectImgStyled = styled(motion.img)<{
  size: number;
}>`
  top: ${({ size }) => (size * 38)/100}px;
  left: ${({ size }) => (size - (size * 28)/100) / 2 - (size * 2)/100}px;
  height: ${({ size }) => (size * 28)/100}px;
  width: ${({ size }) => (size * 28)/100}px;
  position: absolute;
`;

const ObjectInRoadAsset = (props: ObjectInRoadAssetProps) => {
  return (
    <ObjectImgStyled
      size={props.size}
      alt={props.obj.type}
      animate={{
        translateX: props.matrix[2 - props.index][props.obj.y].x * props.size,
        translateY: props.matrix[props.index][props.obj.y].y * props.size * -1,
        scale: props.matrix[2 - props.index][props.obj.y].scale,
        transition: {
          ease: "easeInOut"
        }
      }}
      data-testid="obj-img"
      style={{ scale: 0.18 }}
      onAnimationComplete={props.onAnimationComplete(props.obj, props.index)}
      src={props.obj.type === "rock" ? rockImg : lightningImg}
    />
  );
};

ObjectInRoadAsset.defaultProps = {};

export default ObjectInRoadAsset;
