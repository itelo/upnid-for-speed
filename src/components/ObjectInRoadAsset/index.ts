export { default as ObjectInRoadAsset } from './ObjectInRoadAsset';
