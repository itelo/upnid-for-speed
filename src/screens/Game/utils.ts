export const getRandomInt = (min: number, max: number) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
};

export const generateEmpty = (index: number) => ({
  y: null,
  key: `key_${index}`
});

export const generateNitroOrRock = (index: number) => ({
  y: getRandomInt(0, 2),
  key: `key_${index}`,
  type:
    getRandomInt(0, 1) === 0 ? "lightning" : ("rock" as "rock" | "lightning")
});

export const generateRoad = (size: number, probabilityOfAppearSomething = 5) =>
  new Array(size)
    .fill(0)
    .map((_, i) =>
      Math.random() * 10 > probabilityOfAppearSomething
        ? generateEmpty(i)
        : generateNitroOrRock(i)
    );

export const getNewPos = (
  currentPos: [number, number],
  dir: "left" | "right" | "up" | "down"
): [number, number] => {
  const minPos = 0;
  const maxPos = 2;
  const currentX = currentPos[0];
  const currentY = currentPos[1];
  switch (true) {
    case dir === "left":
      if (currentX - 1 < minPos) {
        return [minPos, currentY];
      } else {
        return [currentX - 1, currentY];
      }
    case dir === "right":
      if (currentX + 1 > maxPos) {
        return [maxPos, currentY];
      } else {
        return [currentX + 1, currentY];
      }
    case dir === "down":
      if (currentY + 1 > maxPos) {
        return [currentX, maxPos];
      } else {
        return [currentX, currentY + 1];
      }
    case dir === "up":
      if (currentY - 1 < minPos) {
        return [currentX, minPos];
      } else {
        return [currentX, currentY - 1];
      }
  }
  return [1, 1];
};


export const matrix = [
  [
    {
      x: -0.045,
      y: -0.3,
      scale: 0.18
    },
    {
      x: 0,
      y: -0.3,
      scale: 0.18
    },
    {
      x: 0.045,
      y: -0.3,
      scale: 0.18
    }
  ],
  [
    {
      x: -0.2,
      y: -0.1,
      scale: 0.55
    },
    {
      x: 0,
      y: -0.1,
      scale: 0.55
    },
    {
      x: 0.2,
      y: -0.1,
      scale: 0.55
    }
  ],
  [
    {
      x: -0.35,
      y: 0,
      scale: 1
    },
    {
      x: 0,
      y: 0,
      scale: 1
    },
    {
      x: 0.35,
      y: 0,
      scale: 1
    }
  ]
];