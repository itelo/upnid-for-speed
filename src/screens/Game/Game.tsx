import React from "react";
import { motion } from "framer-motion";
import styled from "styled-components";
import roadGif from "../../assets/road.gif";
import { useStep, useKeyboard } from "../../hooks";
import CounterScreen from "../CounterScreen/CounterScreen";
import { useNitro } from "../../hooks/useNitro";
import PauseScreen from "../PauseScreen/PauseScreen";
import {
  CarAsset,
  ObjectInRoadAsset,
  Typography,
  Nitro,
  Button
} from "../../components";
import { generateRoad, matrix, getNewPos } from "./utils";
import { SizeContext } from "../../context";

type ArgumentTypes<F extends Function> = F extends (...args: infer A) => any
  ? A
  : never;

type GameProps = {
  onClickExit: ArgumentTypes<typeof Button>[number]["onClick"];
  username: string;
};

const initialCarPos = [1, matrix.length - 1] as [number, number];

const Game = (props: GameProps) => {
  const [carPosition, setT] = React.useState(initialCarPos);
  const [pointsCounter, setPointsCounter] = React.useState(0);
  const nitro = useNitro();

  const { step, setStep, next: nextStep } = useStep(1, 3);

  const [posR, setPosR] = React.useState(generateRoad(20));

  const { size } = React.useContext(SizeContext);

  const handleKeypress = React.useCallback(
    (e: KeyboardEvent) => {
      if (step === 1) {
        if (e.key === "a" || e.key === "ArrowLeft") {
          setT(getNewPos(carPosition, "left"));
        }

        if (e.key === "d" || e.key === "ArrowRight") {
          setT(getNewPos(carPosition, "right"));
        }

        if (e.key === "s" || e.key === "ArrowDown") {
          setT(getNewPos(carPosition, "down"));
        }

        if (e.key === "w" || e.key === "ArrowUp") {
          setT(getNewPos(carPosition, "up"));
        }

        if (e.key === "Escape") {
          setStep(2);
        }

        if (e.code === "Space") {
          nitro.isActive ? nitro.stop() : nitro.active();
        }
      }
    },
    [step, carPosition, setStep, nitro]
  );

  useKeyboard({
    onKeypress: handleKeypress
  });

  React.useEffect(() => {
    const id = setInterval(
      () => {
        if (step === 1) {
          setPointsCounter(x => x + 1);
          setPosR(x => [...x.slice(1, x.length), ...x.slice(0, 1)]);
        }
      },
      nitro.isActive ? 200 : 400
    );

    return () => {
      clearInterval(id);
    };
  }, [step, nitro.isActive]);

  const handleFreshStart = () => {
    setT(initialCarPos);
    setPosR(generateRoad(20));
    setStep(0);
    nitro.reset();
    setPointsCounter(0);
  };

  const handleAnimationComplete = (
    element: {
      y: number;
      type: "rock" | "lightning";
    },
    index: number
  ) => () => {
    if (element.y === carPosition[0] && 2 - index === carPosition[1]) {
      if (element.type === "rock") {
        return setStep(3);
      }
      if (element.type === "lightning") {
        return nitro.charge();
      }
    }
  };

  return (
    <>
      <div style={{ display: "flex", justifyContent: "center" }}>
        <div style={{ position: "relative", height: size, width: size }}>
          <Nitro value={nitro.value} />

          {step === 0 && <CounterScreen onFinish={() => nextStep()} />}
          {step === 2 && (
            <PauseScreen
              onClickContinue={() => setStep(0)}
              title="PAUSE"
              buttonLabel="CONTINUE"
            />
          )}
          {step === 3 && (
            <PauseScreen
              onClickExit={props.onClickExit}
              onClickContinue={handleFreshStart}
              title="YOU DIED"
              message={`${pointsCounter} points`}
              buttonLabel="TRY AGAIN"
              buttonLabel2="Exit"
            />
          )}

          <img
            alt="gif of road"
            src={roadGif}
            style={{
              height: size,
              width: size
            }}
          />
          <div style={{ position: "absolute", top: 25, right: 25 }}>
            <Typography
              family="PressStart2P"
              color="yellow"
              style={{ margin: "0.3rem 1rem", fontSize: "2rem" }}
            >
              {pointsCounter} <span style={{ fontSize: "2rem" }}>points</span>
            </Typography>
          </div>
          <div style={{ position: "absolute", top: 'calc(25px + 3rem)', right: 25 }}>
            <Typography
              family="PressStart2P"
              color="yellow"
              style={{ margin: "0.3rem 1rem", fontSize: "2rem" }}
            >
              {props.username}
            </Typography>
          </div>
          <CarAsset size={size} matrix={matrix} position={carPosition} />
          {posR.slice(0, 3).map((rockPos, i) => {
            if (rockPos.y === null) return null;
            return (
              <ObjectInRoadAsset
                key={rockPos.key}
                matrix={matrix}
                size={size}
                obj={rockPos}
                onAnimationComplete={handleAnimationComplete}
                index={i}
              />
            );
          })}
        </div>
      </div>
    </>
  );
};

Game.defaultProps = {};

export default Game;
