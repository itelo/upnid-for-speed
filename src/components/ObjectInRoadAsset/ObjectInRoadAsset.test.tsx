import React from "react";
import { render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import ObjectInRoadAsset from "./ObjectInRoadAsset";
import { matrix } from "../../screens/Game/utils";
import "jest-styled-components";

describe("<CarAsset />", () => {
  const leftPosition = [0, 0] as [number, number];
  const middlePosition = [1, 0] as [number, number];
  const rightPosition = [2, 0] as [number, number];

  const objLightning = {
    y: 0,
    key: "lightning",
    type: "lightning" as "lightning"
  };

  const objRock = {
    y: 0,
    key: "rock",
    type: "rock" as "rock"
  };

  it("test basic render of lightning", async () => {
    const onAnimationComplete = jest.fn();

    const { getByTestId, container } = render(
      <ObjectInRoadAsset
        index={0}
        size={100}
        matrix={matrix}
        obj={objLightning}
        onAnimationComplete={() => onAnimationComplete()}
      />
    );

    const element = getByTestId("obj-img");

    await waitFor(() =>
      expect(element).toHaveStyle(
        "transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
      )
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("test if the onAnimationComplete is called", async () => {
    const onAnimationComplete = jest.fn();

    const { getByTestId, rerender, container } = render(
      <ObjectInRoadAsset
        index={1}
        size={100}
        matrix={matrix}
        obj={objLightning}
        onAnimationComplete={() => onAnimationComplete()}
      />
    );

    const element = getByTestId("obj-img");

    await waitFor(() =>
      expect(element).toHaveStyle(
        "transform: translateX(-20px) translateY(10px) scale(0.55) translateZ(0);"
      )
    );

    rerender(
      <ObjectInRoadAsset
        index={0}
        size={100}
        matrix={matrix}
        obj={objLightning}
        onAnimationComplete={() => onAnimationComplete()}
      />
    );

    await waitFor(() =>
      expect(element).toHaveStyle(
        "transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
      )
    );

    expect(onAnimationComplete).toHaveBeenCalledTimes(2);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("test basic render of rock", async () => {
    const onAnimationComplete = jest.fn();

    const { getByTestId, container } = render(
      <ObjectInRoadAsset
        index={0}
        size={100}
        matrix={matrix}
        obj={objRock}
        onAnimationComplete={() => onAnimationComplete()}
      />
    );

    const element = getByTestId("obj-img");

    await waitFor(() =>
      expect(element).toHaveStyle(
        "transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
      )
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("test if the onAnimationComplete is called", async () => {
    const onAnimationComplete = jest.fn();

    const { getByTestId, rerender, container } = render(
      <ObjectInRoadAsset
        index={1}
        size={100}
        matrix={matrix}
        obj={objRock}
        onAnimationComplete={() => onAnimationComplete()}
      />
    );

    const element = getByTestId("obj-img");

    await waitFor(() =>
      expect(element).toHaveStyle(
        "transform: translateX(-20px) translateY(10px) scale(0.55) translateZ(0);"
      )
    );

    rerender(
      <ObjectInRoadAsset
        index={0}
        size={100}
        matrix={matrix}
        obj={objRock}
        onAnimationComplete={() => onAnimationComplete()}
      />
    );

    await waitFor(() =>
      expect(element).toHaveStyle(
        "transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
      )
    );

    expect(onAnimationComplete).toHaveBeenCalledTimes(2);
    expect(container.firstChild).toMatchSnapshot();
  });
});
