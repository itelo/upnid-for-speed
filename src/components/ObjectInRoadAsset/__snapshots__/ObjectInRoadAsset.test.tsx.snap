// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarAsset /> test basic render of lightning 1`] = `
.c0 {
  top: 38px;
  left: 34px;
  height: 28px;
  width: 28px;
  position: absolute;
}

<img
  alt="lightning"
  class="c0"
  data-testid="obj-img"
  size="100"
  src="lightning-bolt.png"
  style="transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
/>
`;

exports[`<CarAsset /> test basic render of rock 1`] = `
.c0 {
  top: 38px;
  left: 34px;
  height: 28px;
  width: 28px;
  position: absolute;
}

<img
  alt="rock"
  class="c0"
  data-testid="obj-img"
  size="100"
  src="rock.png"
  style="transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
/>
`;

exports[`<CarAsset /> test if the onAnimationComplete is called 1`] = `
.c0 {
  top: 38px;
  left: 34px;
  height: 28px;
  width: 28px;
  position: absolute;
}

<img
  alt="lightning"
  class="c0"
  data-testid="obj-img"
  size="100"
  src="lightning-bolt.png"
  style="transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
/>
`;

exports[`<CarAsset /> test if the onAnimationComplete is called 2`] = `
.c0 {
  top: 38px;
  left: 34px;
  height: 28px;
  width: 28px;
  position: absolute;
}

<img
  alt="rock"
  class="c0"
  data-testid="obj-img"
  size="100"
  src="rock.png"
  style="transform: translateX(-35px) translateY(30px) scale(1) translateZ(0);"
/>
`;
